### ISSUE TITLE
`MulticlassAccuracy` using lot of memory with many classes

### ISSUE THREAD

**Issue by lucadiliello** [Lightning-AI/metrics#1279]
## 🐛 Bug

After the refactor of the Accuracy package, I am experiencing memory problems in the classification with many classes. Basically, I have to compute the accuracy over 50K classes (output of a Transformer trained with MLM). I just need the “micro” accuracy, so something like `(preds == target).mean()`. I do not understand how I could use `MulticlassAccuracy` to avoid the instantiation of a tensor of 50K*50K integers, which requires about 18GB of GPU memory.

### To Reproduce

Just use `MulticlassAccuracy` with a large number of classes.

#### Code sample

```python
accu = MulticlassAccuracy(num_classes=50000, average='micro')
preds = torch.tensor([[0, 1, 2], [1, 1, 2]])
target = torch.tensor([[0, 1, 1], [2, 2, 1]])
accu(preds, target)
```

### Expected behavior

With `micro` average, it not necessary to instantiate a matrix of `num_classes^2` to store the predictions over each class. The problematic line is the following `https://github.com/Lightning-AI/metrics/blob/84ef54a281af00fe93b98ab89696e393969654a9/src/torchmetrics/functional/classification/stat_scores.py#L405`

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source): pip 0.10.0
- Python & PyTorch Version (e.g., 1.0): 3.9.12
- Any other relevant information such as OS (e.g., Linux): MPS on macOS 12.6 and Ubuntu 20.04 with CUDA 11.6
